### PR TITLE
[ISSUE-642][FEATURE] worker storage manger store user to shuffke key relation and recover from level db

### PR DESCRIPTION
**Pull request by nafiyAix**
### What changes were proposed in this pull request?
storage manger store user to shuffke key relation and recover from level db

### Why are the changes needed?
ensure we could store user's shuffle info even the worker stops.

### What are the items that need reviewer attention?
1. Should we also clean the expired `userShuffleKeys` mapping when running `cleanupExpiredAppDirs()` and `cleanupExpiredShuffleKey()`, or clean it when using this mapping to collect user resource consumption?
2. Do we need to set independent expire policies for `userShuffleKeys` itself like if some user's job haven't send shuffle to current worker for a time then storage manager will remove this user's mapping?

### Related issues.
#642 

### How was this patch tested?

/cc @AngersZhuuuu 

/assign @main-reviewer

### CODE REVIEW

**Comment by nafiyAix** [comment]
Current leveldb file is   
```
private val RECOVERY_FILE_INFOS_FILE_NAME = "fileInfos.ldb"
```
If we change it to other name like `recover.ldb`, will it have potential compatibility issue?

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/storage/StorageManager.scala [hunk 1]
```diff
@@ -166,18 +169,19 @@ final private[worker] class StorageManager(conf: RssConf, workerSource: Abstract
   private val fileInfos =
     new ConcurrentHashMap[String, ConcurrentHashMap[String, FileInfo]]()
   private val RECOVERY_FILE_INFOS_FILE_NAME = "fileInfos.ldb"
-  private var fileInfosDb: DB = null
+  private var db: DB = null
```
Comment (AngersZhuuuu) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/storage/StorageManager.scala:172]: Pls add comment about the storage  sequence of file infos and user shuffle keys here

**Review by AngersZhuuuu — COMMENTED** [review]
Seems missing remove shuffle key in `cleanupExpiredShuffleKey()`

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/storage/StorageManager.scala [hunk 2]
```diff
@@ -49,6 +50,8 @@ final private[worker] class StorageManager(conf: RssConf, workerSource: Abstract
   extends ShuffleRecoverHelper with DeviceObserver with Logging with MemoryTrackerListener {
   // mount point -> filewriter
   val workingDirWriters = new ConcurrentHashMap[File, util.ArrayList[FileWriter]]()
+  // userIdentifier -> shuffleKey
+  val userShuffleKeys = new ConcurrentHashMap[UserIdentifier, util.Set[String]]()
```
Comment (AngersZhuuuu) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/storage/StorageManager.scala:54]: Why no conflicts with current master branch?

File: server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/ShuffleRecoverHelper.java [hunk 3]
```diff
@@ -19,8 +19,12 @@
 
 import java.nio.charset.StandardCharsets;
 
+import com.aliyun.emr.rss.common.protocol.message.ControlMessages;
+import com.aliyun.emr.rss.common.protocol.message.ControlMessages.UserIdentifier;
+
 public abstract class ShuffleRecoverHelper {
   protected String SHUFFLE_KEY_PREFIX = "SHUFFLE-KEY";
+  protected String USER_IDENTIFIER_PREFIX = "USER_IDENTIFIER";
```
Comment (AngersZhuuuu) [inline server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/ShuffleRecoverHelper.java:27]: `USER-IDENTIFIER`?

File: common/src/main/java/com/aliyun/emr/rss/common/util/PbSerDeUtils.java [hunk 4]
```diff
@@ -111,6 +111,19 @@ public static byte[] toPbFileInfoMap(ConcurrentHashMap<String, FileInfo> fileInf
     return builder.build().toByteArray();
   }
 
+  public static Set<String> fromPbShuffleKeys(byte[] data) throws InvalidProtocolBufferException {
```
Comment (AngersZhuuuu) [inline common/src/main/java/com/aliyun/emr/rss/common/util/PbSerDeUtils.java:114]: `fromPbShuffleKeySet `

File: common/src/main/java/com/aliyun/emr/rss/common/util/PbSerDeUtils.java [hunk 5]
```diff
@@ -111,6 +111,19 @@ public static byte[] toPbFileInfoMap(ConcurrentHashMap<String, FileInfo> fileInf
...
+    PbShuffleKeys pbShuffleKeys = PbShuffleKeys.parseFrom(data);
+    Set<String> shuffleKeys = ConcurrentHashMap.newKeySet();
+    shuffleKeys.addAll(pbShuffleKeys.getShuffleKeysList());
+    return shuffleKeys;
+  }
+
+  public static byte[] toPbShuffleKeys(Set<String> shuffleKeys) {
```
Comment (AngersZhuuuu) [inline common/src/main/java/com/aliyun/emr/rss/common/util/PbSerDeUtils.java:121]: `toPbShuffleKeySet`?

**Comment by nafiyAix** [comment]
> Current leveldb file is
> 
> ```
> private val RECOVERY_FILE_INFOS_FILE_NAME = "fileInfos.ldb"
> ```
> 
> If we change it to other name like `recover.ldb`, will it have potential compatibility issue?

FYI: this file name has been changed to recovery.ldb, consider currently may not so many people is using this function, thus we can fix it before it gets tricky.

**Comment by AngersZhuuuu** [comment]
ping @waitinfuture

File: worker/src/main/java/org/apache/celeborn/service/deploy/worker/ShuffleRecoverHelper.java [hunk 6]
```diff
@@ -29,8 +33,22 @@ protected byte[] dbShuffleKey(String shuffleKey) {
...
+    return (USER_IDENTIFIER_PREFIX + ";" + userIdentifier.toString())
+        .getBytes(StandardCharsets.UTF_8);
+  }
+
+  protected UserIdentifier parseDbUserIdentifier(String s) {
+    if (!s.startsWith(USER_IDENTIFIER_PREFIX)) {
+      throw new IllegalArgumentException(
+          "Expected a string starting with " + USER_IDENTIFIER_PREFIX);
+    }
+    return ControlMessages.UserIdentifier$.MODULE$.apply(
```
Comment (AngersZhuuuu) [inline worker/src/main/java/org/apache/celeborn/service/deploy/worker/ShuffleRecoverHelper.java:51]: Can remove `ControlMessage`?

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StorageManager.scala [hunk 7]
```diff
@@ -219,11 +223,46 @@ final private[worker] class StorageManager(conf: RssConf, workerSource: Abstract
...
+  }
+
+  private def reloadAndCleanUserShuffleKeys(db: DB): Unit = {
+    if (db != null) {
+      val itr = db.iterator
+      itr.seek(USER_IDENTIFIER_PREFIX.getBytes(StandardCharsets.UTF_8))
+      while (itr.hasNext) {
+        val entry = itr.next
+        val key = new String(entry.getKey, StandardCharsets.UTF_8)
+        if (key.startsWith(USER_IDENTIFIER_PREFIX)) {
```
Comment (waitinfuture) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StorageManager.scala:242]: Should we break when the condition is false?

**Comment by waitinfuture** [comment]
> 

I think we should cleanup userShuffleKeys whenever shuffle data is cleaned, or the map will grow infinitely.

**Comment by waitinfuture** [comment]
BTW, do we need the userShuffleKeys? Can we add UserIdentifier into FileInfo?

**Comment by AngersZhuuuu** [comment]
> > 
> 
> I think we should cleanup userShuffleKeys whenever shuffle data is cleaned, or the map will grow infinitely.

Seems I forgot to submit the review record of the same comment..

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StorageManager.scala (hunk 7 above)
Comment (nafiyAix) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StorageManager.scala:242]: Agree, seems the iterator's keys are sorted in lexicographic order by default.

**Comment by nafiyAix** [comment]
> BTW, do we need the userShuffleKeys? Can we add UserIdentifier into FileInfo instead?

Agree, will move it.

**Comment by nafiyAix** [comment]
> BTW, do we need the userShuffleKeys? Can we add UserIdentifier into FileInfo instead?

Do we need to create a new issue for this change? Because seems like there will be many chain reactions if we put UserIdentifier into FileInfo, especially those logic related with PartitionFileSorter.openStream()
```
public FileInfo openStream(
      String shuffleKey, String fileName, FileInfo fileInfo, int startMapIndex, int endMapIndex)
      throws IOException {
```

**Comment by waitinfuture** [comment]
> > BTW, do we need the userShuffleKeys? Can we add UserIdentifier into FileInfo instead?
> 
> Do we need to create a new issue for this change? Because seems like there will be many chain reactions if we put UserIdentifier into FileInfo as its field, especially those logic related with PartitionFileSorter.openStream() cc: @waitinfuture
> 
> ```
> public FileInfo openStream(
>       String shuffleKey, String fileName, FileInfo fileInfo, int startMapIndex, int endMapIndex)
>       throws IOException {
> ```

Agree, we should further consider it.

**Review by waitinfuture — APPROVED** [review]
LGTM, thanks!

**Pull request merged**